### PR TITLE
python312Packages.qtile-bonsai: init at 0.4.0

### PR DESCRIPTION
**Pull request by Gurjaka**
## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Sigmanificient** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356700`

---
### `x86_64-linux`

File: pkgs/development/python-modules/qtile-bonsai/default.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  cffi,
+  strenum,
+  psutil,
+  xcffib,
+  pdm-backend,
+  pytestCheckHook,
+  pyside6,
+  pyvirtualdisplay,
+  qtile,
+  extraPackages ? [ ],
```
Comment (pluiedev) [inline pkgs/development/python-modules/qtile-bonsai/default.nix:17]: Why is this necessary?

File: pkgs/development/python-modules/qtile-bonsai/default.nix [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    pdm-backend
+  ];
+
+  dependencies = extraPackages ++ [
+    strenum
+    psutil
+    pyside6
+    pyvirtualdisplay
+    setuptools
+    setuptools-scm
```
Comment (pluiedev) [inline pkgs/development/python-modules/qtile-bonsai/default.nix:43]: These should be in `build-system` and not be normal dependencies usually

File: pkgs/development/python-modules/qtile-bonsai/default.nix [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    (cairocffi.override { withXcffib = true; })
+    cffi
+    xcffib
+    qtile
+  ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    qtile
```
Comment (pluiedev) [inline pkgs/development/python-modules/qtile-bonsai/default.nix:52]: I don't think this is necessary if it's already listed as a dependency

File: pkgs/development/python-modules/qtile-bonsai/default.nix [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    export HOME=$(mktemp -d)
+  '';
+
+  disabledTestPaths = [
+    # Needs a running DBUS
+    "tests/integration/test_layout.py"
+    "tests/integration/test_widget.py"
+  ];
+
+  pythonImportCheck = [ "qtile-bonsai" ];
```
Comment (pluiedev) [inline pkgs/development/python-modules/qtile-bonsai/default.nix:65]: Pretty sure it's snake_case (I'm not even sure if Python allows hyphens in module names)
```suggestion
  pythonImportCheck = [ "qtile_bonsai" ];
```

**Review by pluiedev — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/qtile-bonsai/default.nix (hunk 2 above)
Comment (Gurjaka) [reply pkgs/development/python-modules/qtile-bonsai/default.nix:43]: you are right just removed them and works fine

**Review by pluiedev — APPROVED** [review]
Commit & PR should be retitled to `python312Packages.qtile-bonsai: init at 0.4.0`, and the package should be added to `pkgs/top-level/python-packages.nix`

**Comment by Gurjaka** [comment]
> Commit & PR should be retitled to `python312Packages.qtile-bonsai: init at 0.4.0`, and the package should be added to `pkgs/top-level/python-packages.nix`

python312Packages or python3Packages?

**Comment by pluiedev** [comment]
> > Commit & PR should be retitled to `python312Packages.qtile-bonsai: init at 0.4.0`, and the package should be added to `pkgs/top-level/python-packages.nix`

python312Packages

**Comment by Gurjaka** [comment]
> > python312Packages or python3Packages?
> 
> python312Packages

May i ask the difference between two? I am new to all this

**Comment by pluiedev** [comment]
`python3Packages` refers to the default package set using the default Python interpreter (currently 3.12) while `python312Packages` specifically refers to Python 3.12. It's preferred to specifically indicate the used Python version in the commit and PR title, because it shows reviewers exactly which version of Python you're using and makes sure that they will use the same version to test your package.

**Comment by Gurjaka** [comment]
> `python3Packages` refers to the default package set using the default Python interpreter (currently 3.12) while `python312Packages` specifically refers to Python 3.12. It's preferred to specifically indicate the used Python version in the commit and PR title, because it shows reviewers exactly which version of Python you're using and makes sure that they will use the same version to test your package. It also allows @ofborg to build your package as a part of CI.
> 
> Quoting https://nixos.org/manual/nixpkgs/unstable/#contributing-guidelines (emphasis mine):
> 
> > Commit names of Python libraries should reflect that they are Python libraries, so write for example python311Packages.numpy: 1.11 -> 1.12. **It is highly recommended to specify the current default version to enable automatic build by ofborg.** Note that pythonPackages is an alias for python27Packages.

Thanks for detailed answer, I changed everything, and I guess we are ready to be merged!

**Comment by Gurjaka** [comment]
> Well, I don't have committer access, but I can forward this to the PRs already reviewed Discourse thread for merging

I have to resolve python-packages.nix first

**Comment by Gurjaka** [comment]
> > Well, I don't have committer access, but I can forward this to the PRs already reviewed Discourse thread for merging
> 
> I have to resolve python-packages.nix first

alright its done

**Review by FliegendeWurst — APPROVED** [review]

File: pkgs/development/python-modules/qtile-bonsai/default.nix [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  ];
+
+  nativeCheckInputs = [
+    pyside6
+    pyvirtualdisplay
+    (cairocffi.override { withXcffib = true; })
+    cffi
+    xcffib
+    qtile
+  ];
```
Comment (natsukium) [inline pkgs/development/python-modules/qtile-bonsai/default.nix:44]: Please add `pytestCheckHook` to nativeCheckinputs or provide `checkPhase` to enable tests.
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pytestcheckhook-using-pytestcheckhook

**Review by natsukium — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/qtile-bonsai/default.nix (hunk 5 above)
Comment (Gurjaka) [reply pkgs/development/python-modules/qtile-bonsai/default.nix]: done

**Review by Sigmanificient — APPROVED** [review]

**Review by natsukium — APPROVED** [review]
LGTM, thanks.

**Comment by natsukium** [comment]
Welcome to nixpkgs🎉
